Support offset in composite aggs

nik9000 opened the pull request:
Adds support for the `offset` parameter to the `date_histogram` source
of composite aggs. The `offset` parameter is supported by the normal
`date_histogram` aggregation and is useful for folks that need to
measure things from, say, 6am one day to 6am the next day.

This is implemented by creating a new `Rounding` that knows how to
handle offsets and delegates to other rounding implementations. That
implementation doesn't fully implement the `Rounding` contract, namely
`nextRoundingValue`. That method isn't used by composite aggs so I can't
be sure that any implementation that I add will be correct. I propose to
leave it throwing `UnsupportedOperationException` until I need it.

Closes #48757

nik9000 committed:
Support offset in composite aggs

Adds support for the `offset` parameter to the `date_histogram` source
of composite aggs. The `offset` parameter is supported by the normal
`date_histogram` aggregation and is useful for folks that need to
measure things from, say, 6am one day to 6am the next day.

This is implemented by creating a new `Rounding` that knows how to
handle offsets and delegates to other rounding implementations. That
implementation doesn't fully implement the `Rounding` contract, namely
`nextRoundingValue`. That method isn't used by composite aggs so I can't
be sure that any implementation that I add will be correct. I propose to
leave it throwing `UnsupportedOperationException` until I need it.

Closes #48757

elasticmachine commented:
Pinging @elastic/es-docs (>docs)

elasticmachine commented:
Pinging @elastic/es-analytics-geo (:Analytics/Aggregations)

nik9000 commented:
Please have a close look at this one! It has wire changes and it has been more than a year since I made any of those....

nik9000 commented on server/src/main/java/org/elasticsearch/common/Rounding.java:
```diff
@@ -556,19 +567,73 @@ public boolean equals(Object obj) {
+
+        @Override
+        public void innerWriteTo(StreamOutput out) throws IOException {
+            if (out.getVersion().before(Version.V_8_0_0)) {
```
We don't actually serialize this `Rounding` but I believe I've got this bit right so I kept it. I'd like to move the other `offset` code over to this Rounding which means that we will *eventually* serialize it. At that point we'll get exhaustive tests for this code.

nik9000 replied:
Modulo the version check which I discuss below.

nik9000 commented on server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/DateHistogramValuesSourceBuilder.java:
```diff
@@ -80,12 +91,18 @@ protected DateHistogramValuesSourceBuilder(StreamInput in) throws IOException {
         super(in);
         dateHistogramInterval = new DateIntervalWrapper(in);
         timeZone = in.readOptionalZoneId();
+        if (in.getVersion().after(Version.V_8_0_0)) {
```
I *think* this is the way to do this, but it has been a long time. I want to land this change in master and 7.x, but these tests have no change of passing the bwc tests until I backport it.

Now that I think about it, maybe *this* version check should be `V_7_6_0` actually. Help!

nik9000 replied:
The tests caught an error - this at least be `onOrAfter` but I'll switch it to `V_7_6_0` while I'm there.

nik9000 commented on docs/reference/aggregations/bucket/composite-aggregation.asciidoc:
```diff
@@ -287,6 +287,72 @@ Time zones may either be specified as an ISO 8601 UTC offset (e.g. `+01:00` or
 
+*Offset*
+
+include::datehistogram-aggregation.asciidoc[tag=offset-explanation]
```
I figured because the option is the same I should just include the docs from the normal date histogram.

nik9000 commented:
Hurray! Integration test failure! Will debug. Good robots, catching things.

nik9000 committed:
version checks

nik9000 commented:
@elasticmachine, run elasticsearch-ci/2

nik9000 commented:
The default distro failure is due to the thing not being backported. I'll find a way to get this working.

nik9000 committed:
use 8.0.0

nik9000 committed:
Revert "use 8.0.0"

This reverts commit 95da5cb0db5008174298d6ff3980246b2426f209.

nik9000 committed:
Disable bwc tests

nik9000 commented:
> The default distro failure is due to the thing not being backported. I'll find a way to get this working.

I believe I've got this sorted thanks to @DaveCTurner and @cbuescher.

polyfractal commented on server/src/test/java/org/elasticsearch/common/RoundingTests.java:
```diff
@@ -195,6 +195,16 @@ public void testTimeUnitRoundingDST() {
         assertThat(tzRounding_chg.round(time("2014-11-02T06:01:01", chg)), isDate(time("2014-11-02T06:00:00", chg), chg));
     }
 
```
Hmm, do you know if we have any serialization tests for `Rounding`?  I was looking to see if we test that the `id()` bytes are correct and don't accidentally change (similar to what we do with `AbstractWriteableEnumTestCase` enum tests)... but couldn't find any serialization tests at all.

If we have them somewhere, let's add a test for the id byte.  If not, probably too much to add to this PR but we should file a ticket so we don't forget to add some tests... makes me uneasy that such a widespread class doesn't have serialization tests :)

nik9000 replied:
I believe we end up testing serialization as part of testing things like extended bounds bucket response. Adding a unit test for just this class's serialization makes sense to me though. I figured I'd wait until I used it in a context where we serialized it but since I'm writing the serialization code now I probably ought to write the test now.

nik9000 replied:
I pushed a wire test case.

polyfractal commented on server/src/test/java/org/elasticsearch/common/RoundingTests.java:
```diff
@@ -195,6 +195,16 @@ public void testTimeUnitRoundingDST() {
         assertThat(tzRounding_chg.round(time("2014-11-02T06:01:01", chg)), isDate(time("2014-11-02T06:00:00", chg), chg));
     }
 
+    public void testOffsetRounding() {
```
Let's add another test that does negative offsets too?

nik9000 replied:
Sure!

nik9000 replied:
I pushed a test with negative offsets.

polyfractal reviewed (commented):
Left a few comments/questions, but I think this looks good.  Main question was around the new Rounding class, most of the rest is cosmetic :)

nik9000 commented on server/src/main/java/org/elasticsearch/common/Rounding.java:
```diff
@@ -556,19 +567,73 @@ public boolean equals(Object obj) {
         }
     }
 
+    static class OffsetRounding extends Rounding {
```
Yeah, I did intend to migrate the rest over of the uses of `offset` over to it.

I added javadoc on the public interface to the class which is what the rest of the classes in this file do. Do you think I should duplicate it onto this one? Or add something maybe?

polyfractal replied:
Ah, missed that it was documented on the public method.  Since the rest of the class does it that way, fine with me :)

Thanks for the explanation!

nik9000 commented on server/src/main/java/org/elasticsearch/common/Rounding.java:
```diff
@@ -556,19 +567,73 @@ public boolean equals(Object obj) {
+
+        OffsetRounding(StreamInput in) throws IOException {
+            delegate = Rounding.read(in);
+            offset = in.readLong();
```
In other places offsets are written as long and I just copied it. But I'd be fine with zlong.

nik9000 replied:
Pushed the switch to zlong.

nik9000 commented on server/src/main/java/org/elasticsearch/common/Rounding.java:
```diff
@@ -556,19 +567,73 @@ public boolean equals(Object obj) {
             case TimeUnitRounding.ID:
-                rounding = new TimeUnitRounding(in);
-                break;
+                return new TimeUnitRounding(in);
```
I'm not sure everyone would consider that cleaner, but I sure do!

nik9000 committed:
Merge branch 'master' into composite_date_offset

nik9000 committed:
Rework public interface and add wire tests

The serialization was subtly wrong, particularly in the Eire time zone.

nik9000 committed:
zlong

nik9000 commented:
@polyfractal, I've pushed everything you asked for.

nik9000 commented:
I reworked the serialization a bit, moving the write methods so they are near the read methods and replacing `writeString(zone.getId())` with `writeZoneId(zone)`. More importantly, I replaced `DateUtil.of(readString())` with `readZoneId`. The former changes the zone on load, making the hashcode and equals inconsistent.

nik9000 commented on server/src/main/java/org/elasticsearch/common/Rounding.java:
```diff
@@ -418,24 +434,16 @@ public boolean equals(Object obj) {
             }
             TimeUnitRounding other = (TimeUnitRounding) obj;
-            return Objects.equals(unit, other.unit) && Objects.equals(timeZone, other.timeZone);
+            return unit == other.unit && timeZone.equals(other.timeZone);
```
I'll revert this change. It snuck in because I thought the comparison was wrong.

nik9000 committed:
Revert equals

nik9000 commented on server/src/main/java/org/elasticsearch/common/Rounding.java:
```diff
@@ -552,23 +560,88 @@ public boolean equals(Object obj) {
+
+        @Override
+        public String toString() {
+            return "Rounding[" + interval + " in " + timeZone + "]";
```
The `toString` implementations weren't consistent which made reading the error messages hard. So I changed them.

nik9000 commented:
Test failure is real!

nik9000 committed:
Fix up test case

Looks like I never ran it!

nik9000 commented:
> Test failure is real!

Because I never ran I locally. I thought I had, but I must have made a mistake.

nik9000 committed:
Disable some more bwc tests for the backport

nik9000 commented:
OK! Tests look right now.

polyfractal commented on server/src/test/java/org/elasticsearch/common/RoundingWireTests.java:
```diff
@@ -0,0 +1,51 @@
+import org.elasticsearch.common.unit.TimeValue;
+import org.elasticsearch.test.AbstractWireSerializingTestCase;
+
+public class RoundingWireTests extends AbstractWireSerializingTestCase<Rounding> {
```
I'm a nervous nellie... would it be possible to explicitly test the "short" deprecated TZs from DateUtils, just to make sure we aren't breaking them?

nik9000 replied:
So I'm not 100% sure I can or should. Like, I can write a test that uses one of the short IDs, but I'm not super familiar with how to use them properly. The randomized testing doesn't like to generate them, presumably because they don't work on all jvms. I know one of them didn't work for me locally.

The class itself only takes ZoneId, not string. So it'd be fairly difficult to get one of those deprecated zone ids into it.

polyfractal replied:
Ah, I see.  And tracing backwards, DateHisto parses the input string with `ZoneId.of()` which should reject these deprecated short IDs as well.

And the only other potential problem area is Rollup (which stored some short TZs before the javatime switch), but we have an alias lookup map for that use-case.

Righto, carry on :+1:

polyfractal reviewed (approved):
Asked for one more test re: deprecated short TZs, but otherwise LGTM :)

nik9000 committed:
Merge branch 'master' into composite_date_offset

jimczi commented on server/src/main/java/org/elasticsearch/common/Rounding.java:
```diff
@@ -554,21 +562,86 @@ public boolean equals(Object obj) {
+        @Override
+        public void innerWriteTo(StreamOutput out) throws IOException {
+            if (out.getVersion().before(Version.V_7_6_0)) {
+                throw new IllegalArgumentException("Offset rounding not supported before 8.0.0");
```
should the message be before 7.6.0 ?

nik9000 replied:
Yeah. I'll fix it.

jimczi reviewed (approved):
LGTM, for next time I think it's ok to just disable all bwc tests and adapt the version in master until the change is backported to 7x. That's how we handle these breaking changes nowadays but you already adapted all the tests so that's useless now.

nik9000 commented on rest-api-spec/src/main/resources/rest-api-spec/test/search.aggregation/230_composite.yml:
```diff
@@ -367,6 +367,40 @@ setup:
+
+  - do:
+      search:
+        rest_total_hits_as_int: true
```
Sure!

nik9000 committed:
Fix from review

nik9000 commented:
Thanks for the reviews @polyfractal and @jimczi !
